File: src/main/java/com/jaffa/rpc/lib/serialization/JavaSerializer.java
package com.jaffa.rpc.lib.serialization;

import lombok.extern.slf4j.Slf4j;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.ObjectInput;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;

@Slf4j
@SuppressWarnings("squid:S1168")
public final class JavaSerializer implements ObjectSerializer {

    @Override
    public byte[] serialize(Object obj) {
        try (ByteArrayOutputStream bos = new ByteArrayOutputStream()) {
            ObjectOutputStream out;
            out = new ObjectOutputStream(bos);
            out.writeObject(obj);
            out.flush();
            return bos.toByteArray();
        } catch (IOException ioException) {
            log.error("Exception while object Java serialization", ioException);
        }
        return null;
    }

    @Override
    public byte[] serializeWithClass(Object obj) {
        return serialize(obj);
    }

    @Override
    public Object deserializeWithClass(byte[] serialized) {
        ByteArrayInputStream bis = new ByteArrayInputStream(serialized);
        try (ObjectInput in = new ObjectInputStream(bis)) {
            return in.readObject();
        } catch (IOException | ClassNotFoundException exception) {
            log.error("Exception while object Java deserialization", exception);
        }
        return null;
    }

    @SuppressWarnings("unchecked")
    @Override
    public <T> T deserialize(byte[] serialized, Class<T> clazz) {
        ByteArrayInputStream bis = new ByteArrayInputStream(serialized);
        try (ObjectInput in = new ObjectInputStream(bis)) {
            return (T) in.readObject();
        } catch (IOException | ClassNotFoundException | ClassCastException exception) {
            log.error("Exception while object Java deserialization", exception);
        }
        return null;
    }
}

File: src/main/kotlin/com/jaffa/rpc/lib/serialization/JavaSerializer.kt
package com.jaffa.rpc.lib.serialization

import com.jaffa.rpc.lib.exception.JaffaRpcExecutionException
import org.slf4j.LoggerFactory
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.ObjectInputStream
import java.io.ObjectOutputStream

class JavaSerializer : ObjectSerializer {

    private val log = LoggerFactory.getLogger(JavaSerializer::class.java)

    companion object {
        const val ERROR_DESERIALIZATION_MESSAGE = "Exception while object Java deserialization"
        const val ERROR_SERIALIZATION_MESSAGE = "Exception while object Java serialization"
    }

    override fun serialize(obj: Any?): ByteArray? {
        try {
            ByteArrayOutputStream().use { bos ->
                val out = ObjectOutputStream(bos)
                out.writeObject(obj)
                out.flush()
                return bos.toByteArray()
            }
        } catch (exception: Exception) {
            throw JaffaRpcExecutionException(exception)
        }
    }

    override fun serializeWithClass(obj: Any?): ByteArray? {
        return serialize(obj)
    }

    override fun deserializeWithClass(serialized: ByteArray?): Any? {
        val bis = ByteArrayInputStream(serialized)
        try {
            ObjectInputStream(bis).use { `in` -> return `in`.readObject() }
        } catch (exception: Exception) {
            throw JaffaRpcExecutionException(exception)
        }
    }

    override fun <T> deserialize(serialized: ByteArray?, clazz: Class<T>?): T? {
        val bis = ByteArrayInputStream(serialized)
        try {
            ObjectInputStream(bis).use { `in` -> return `in`.readObject() as T }
        } catch (exception: Exception) {
            throw JaffaRpcExecutionException(exception)
        }
    }
}